File: android/app/src/main/java/io/github/manhvu1212/tallyo/ui/newsession/NewSessionScreen.kt
package io.github.manhvu1212.tallyo.ui.newsession

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import io.github.manhvu1212.tallyo.R
import io.github.manhvu1212.tallyo.ui.LocalAppContainer
import io.github.manhvu1212.tallyo.ui.components.Ime
import io.github.manhvu1212.tallyo.ui.components.PrimaryButton
import io.github.manhvu1212.tallyo.ui.components.TallyoTextField
import io.github.manhvu1212.tallyo.ui.components.TopBar
import io.github.manhvu1212.tallyo.ui.theme.TallyoColors

@OptIn(ExperimentalLayoutApi::class)
@Composable
fun NewSessionScreen(
    onBack: () -> Unit,
    onCreated: (String) -> Unit,
) {
    val container = LocalAppContainer.current
    val vm: NewSessionViewModel = viewModel(factory = NewSessionViewModel.factory(container.repository))
    val customGames by vm.customGames.collectAsStateWithLifecycle()

    val context = LocalContext.current
    val keyboard = LocalSoftwareKeyboardController.current
    var name by remember { mutableStateOf("") }
    var selectedGameKey by remember { mutableStateOf("tien_len") }
    var customGameName by remember { mutableStateOf("") }
    var playerInput by remember { mutableStateOf("") }
    var players by remember { mutableStateOf(emptyList<String>()) }
    var zeroSum by remember { mutableStateOf(true) }
    var duplicateName by remember { mutableStateOf<String?>(null) }
    var notEnough by remember { mutableStateOf(false) }
    val playerFocus = remember { FocusRequester() }

    val canCreate = players.size >= 2

    val addPlayer = lambda@{
        val trimmed = playerInput.trim()
        if (trimmed.isEmpty()) return@lambda
        if (players.contains(trimmed)) {
            duplicateName = trimmed
            return@lambda
        }
        players = players + trimmed
        playerInput = ""
    }

    val submit = {
        if (!canCreate) {
            notEnough = true
        } else {
            val selectedGameItem = PRESET_GAMES.firstOrNull { it.key == selectedGameKey }
            val isCustom = selectedGameKey == "custom" || selectedGameKey.startsWith("custom_saved:")
            val gameNameToSend = if (selectedGameKey == "custom") {
                customGameName.trim().ifEmpty { context.getString(R.string.game_custom) }
            } else if (selectedGameKey.startsWith("custom_saved:")) {
                selectedGameKey.removePrefix("custom_saved:")
            } else {
                selectedGameItem?.let { context.getString(it.resourceId) } ?: ""
            }

            val defaultSessionName = if (isCustom) {
                gameNameToSend
            } else if (selectedGameItem != null) {
                context.getString(selectedGameItem.resourceId)
            } else {
                context.getString(R.string.new_session_default_name)
            }

            vm.create(
                name = name,
                game = gameNameToSend,
                playerNames = players,
                zeroSum = zeroSum,
                defaultName = defaultSessionName,
                isCustomGame = isCustom && gameNameToSend.isNotBlank() && gameNameToSend != context.getString(R.string.game_custom),
                onCreated = onCreated,
            )
        }
    }

    Column(Modifier.fillMaxSize().background(TallyoColors.Bg)) {
        TopBar(title = stringResource(R.string.new_title), onBack = onBack)
        Column(
            Modifier
                .weight(1f)
                .imePadding()
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
        ) {
            Section(title = stringResource(R.string.new_game_label)) {
                FlowRow(
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                    modifier = Modifier.fillMaxWidth()
                ) {
                    // 1. Preset games (except custom)
                    PRESET_GAMES.filter { it.key != "custom" }.forEach { gameItem ->
                        val isSelected = selectedGameKey == gameItem.key
                        Row(
                            modifier = Modifier
                                .clip(RoundedCornerShape(999.dp))
                                .background(if (isSelected) TallyoColors.PrimaryTintBg else TallyoColors.SurfaceAlt)
                                .border(
                                    1.dp,
                                    if (isSelected) TallyoColors.Primary else TallyoColors.Border,
                                    RoundedCornerShape(999.dp)
                                )
                                .clickable {
                                    selectedGameKey = gameItem.key
                                    zeroSum = gameItem.defaultZeroSum
                                }
                                .padding(horizontal = 16.dp, vertical = 8.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Text(
                                stringResource(gameItem.resourceId),
                                color = if (isSelected) TallyoColors.Primary else TallyoColors.Text,
                                fontSize = 13.sp,
                                fontWeight = FontWeight.Medium
                            )
                        }
                    }

                    // 2. Saved custom games
                    customGames.forEach { customGame ->
                        val key = "custom_saved:${customGame.name}"
                        val isSelected = selectedGameKey == key
                        Row(
                            modifier = Modifier
                                .clip(RoundedCornerShape(999.dp))
                                .background(if (isSelected) TallyoColors.PrimaryTintBg else TallyoColors.SurfaceAlt)
                                .border(
                                    1.dp,
                                    if (isSelected) TallyoColors.Primary else TallyoColors.Border,
                                    RoundedCornerShape(999.dp)
                                )
                                .clickable {
                                    selectedGameKey = key
                                    zeroSum = customGame.defaultZeroSum
                                }
                                .padding(start = 16.dp, end = 8.dp, top = 4.dp, bottom = 4.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Text(
                                text = customGame.name,
                                color = if (isSelected) TallyoColors.Primary else TallyoColors.Text,
                                fontSize = 13.sp,
                                fontWeight = FontWeight.Medium
                            )
                            Spacer(Modifier.size(4.dp))
                            Box(
                                modifier = Modifier
                                    .clip(CircleShape)
                                    .clickable {
                                        vm.deleteCustomGame(customGame.name)
                                        if (selectedGameKey == key) {
                                            selectedGameKey = "tien_len"
                                        }
                                    }
                                    .padding(4.dp)
                            ) {
                                Text(
                                    text = "×",
                                    color = if (isSelected) TallyoColors.Primary else TallyoColors.TextMuted,
                                    fontSize = 16.sp,
                                    fontWeight = FontWeight.Bold
                                )
                            }
                        }
                    }

                    // 3. Preset "custom" game item
                    PRESET_GAMES.firstOrNull { it.key == "custom" }?.let { gameItem ->
                        val isSelected = selectedGameKey == gameItem.key
                        Row(
                            modifier = Modifier
                                .clip(RoundedCornerShape(999.dp))
                                .background(if (isSelected) TallyoColors.PrimaryTintBg else TallyoColors.SurfaceAlt)
                                .border(
                                    1.dp,
                                    if (isSelected) TallyoColors.Primary else TallyoColors.Border,
                                    RoundedCornerShape(999.dp)
                                )
                                .clickable {
                                    selectedGameKey = gameItem.key
                                    zeroSum = gameItem.defaultZeroSum
                                }
                                .padding(horizontal = 16.dp, vertical = 8.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Text(
                                stringResource(gameItem.resourceId),
                                color = if (isSelected) TallyoColors.Primary else TallyoColors.Text,
                                fontSize = 13.sp,
                                fontWeight = FontWeight.Medium
                            )
                        }
                    }
                }
                if (selectedGameKey == "custom") {
                    Spacer(Modifier.height(8.dp))
                    TallyoTextField(
                        value = customGameName,
                        onValueChange = { customGameName = it },
                        placeholder = stringResource(R.string.new_game_custom_placeholder),
                        keyboardOptions = Ime.Next,
                        modifier = Modifier.fillMaxWidth(),
                    )
                }
            }

            Section(title = stringResource(R.string.new_name_label)) {
                TallyoTextField(
                    value = name,
                    onValueChange = { name = it },
                    placeholder = stringResource(R.string.new_name_placeholder),
                    keyboardOptions = Ime.Next,
                    keyboardActions = KeyboardActions(onNext = { playerFocus.requestFocus() }),
                    modifier = Modifier.fillMaxWidth(),
                )
            }

            Section(
                title = stringResource(R.string.new_players_label),
                hint = stringResource(R.string.new_players_hint),
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    TallyoTextField(
                        value = playerInput,
                        onValueChange = { playerInput = it },
                        placeholder = stringResource(R.string.new_players_placeholder),
                        keyboardOptions = Ime.DoneWords,
                        // Handle both Done and Next: some IMEs keep the previous
                        // field's Next state even after focus moves to a Done field.
                        keyboardActions = KeyboardActions(
                            onDone = {
                                if (playerInput.trim().isEmpty()) keyboard?.hide() else addPlayer()
                            },
                            onNext = {
                                if (playerInput.trim().isEmpty()) keyboard?.hide() else addPlayer()
                            },
                        ),
                        focusRequester = playerFocus,
                        modifier = Modifier.weight(1f),
                    )
                    Spacer(Modifier.size(8.dp))
                    PrimaryButton(
                        label = stringResource(R.string.new_players_add),
                        onClick = addPlayer,
                    )
                }
                Spacer(Modifier.height(12.dp))
                if (players.isEmpty()) {
                    Text(
                        stringResource(R.string.new_players_empty),
                        color = TallyoColors.TextMuted,
                        fontSize = 11.sp,
                    )
                } else {
                    FlowRow(
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        players.forEach { p ->
                            Row(
                                modifier = Modifier
                                    .clip(RoundedCornerShape(999.dp))
                                    .background(TallyoColors.SurfaceAlt)
                                    .border(1.dp, TallyoColors.Border, RoundedCornerShape(999.dp))
                                    .clickable { players = players.filterNot { it == p } }
                                    .padding(horizontal = 12.dp, vertical = 8.dp),
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                Text(p, color = TallyoColors.Text, fontSize = 13.sp, fontWeight = FontWeight.Medium)
                                Text("  ×", color = TallyoColors.TextMuted, fontSize = 16.sp)
                            }
                        }
                    }
                }
            }

            Section(title = stringResource(R.string.new_options_label)) {
                val shape = RoundedCornerShape(10.dp)
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(shape)
                        .background(if (zeroSum) TallyoColors.PrimaryTintBg else TallyoColors.Surface)
                        .border(
                            1.dp,
                            if (zeroSum) TallyoColors.Primary else TallyoColors.Border,
                            shape,
                        )
                        .clickable { zeroSum = !zeroSum }
                        .padding(12.dp),
                ) {
                    Column {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            CheckBox(checked = zeroSum)
                            Spacer(Modifier.size(8.dp))
                            Text(
                                stringResource(R.string.new_options_zero_sum_title),
                                color = TallyoColors.Text,
                                fontSize = 16.sp,
                                fontWeight = FontWeight.SemiBold,
                            )
                        }
                        Text(
                            stringResource(R.string.new_options_zero_sum_desc),
                            color = TallyoColors.TextMuted,
                            fontSize = 13.sp,
                            modifier = Modifier.padding(top = 6.dp, start = 28.dp),
                            lineHeight = 18.sp,
                        )
                    }
                }
            }

            PrimaryButton(
                label = stringResource(R.string.new_submit),
                onClick = submit,
                enabled = canCreate,
                modifier = Modifier.fillMaxWidth().padding(top = 16.dp),
            )
            Spacer(Modifier.height(48.dp))
        }
    }

    duplicateName?.let { dup ->
        AlertDialog(
            onDismissRequest = { duplicateName = null },
            title = { Text(stringResource(R.string.new_players_duplicate_title), color = TallyoColors.Text) },
            text = {
                Text(
                    stringResource(R.string.new_players_duplicate_message, dup),
                    color = TallyoColors.TextMuted,
                )
            },
            containerColor = TallyoColors.Surface,
            confirmButton = {
                TextButton(onClick = { duplicateName = null }) {
                    Text("OK", color = TallyoColors.Primary)
                }
            },
        )
    }

    if (notEnough) {
        AlertDialog(
            onDismissRequest = { notEnough = false },
            title = { Text(stringResource(R.string.new_min_players), color = TallyoColors.Text) },
            containerColor = TallyoColors.Surface,
            confirmButton = {
                TextButton(onClick = { notEnough = false }) { Text("OK", color = TallyoColors.Primary) }
            },
        )
    }
}

@Composable
private fun Section(
    title: String,
    hint: String? = null,
    content: @Composable () -> Unit,
) {
    Column(Modifier.padding(bottom = 24.dp)) {
        Text(title, color = TallyoColors.Text, fontSize = 18.sp, fontWeight = FontWeight.SemiBold)
        if (hint != null) Text(hint, color = TallyoColors.TextMuted, fontSize = 11.sp, modifier = Modifier.padding(top = 2.dp))
        Spacer(Modifier.height(8.dp))
        content()
    }
}

@Composable
private fun CheckBox(checked: Boolean) {
    val color = if (checked) TallyoColors.Primary else TallyoColors.Border
    Box(
        modifier = Modifier
            .size(20.dp)
            .clip(RoundedCornerShape(4.dp))
            .background(if (checked) TallyoColors.Primary else androidx.compose.ui.graphics.Color.Transparent)
            .border(2.dp, color, RoundedCornerShape(4.dp)),
        contentAlignment = Alignment.Center,
    ) {
        if (checked) Text("✓", color = androidx.compose.ui.graphics.Color.White, fontSize = 13.sp, fontWeight = FontWeight.Bold)
    }
}

private data class GameItem(
    val key: String,
    val resourceId: Int,
    val defaultZeroSum: Boolean,
)

private val PRESET_GAMES = listOf(
    GameItem("tien_len", R.string.game_tien_len, true),
    GameItem("phom", R.string.game_phom, true),
    GameItem("mau_binh", R.string.game_mau_binh, true),
    GameItem("poker", R.string.game_poker, true),
    GameItem("sam", R.string.game_sam, true),
    GameItem("ludo", R.string.game_ludo, false),
    GameItem("monopoly", R.string.game_monopoly, false),
    GameItem("custom", R.string.game_custom, false),
)

File: ios/Tallyo/UI/Screens/NewSessionScreen.swift
import SwiftUI
import SwiftData

struct PresetGameItem: Identifiable {
    let id: String
    let localizationKey: String
    let defaultZeroSum: Bool
}

let PRESET_GAMES = [
    PresetGameItem(id: "tien_len", localizationKey: "game_tien_len", defaultZeroSum: true),
    PresetGameItem(id: "phom", localizationKey: "game_phom", defaultZeroSum: true),
    PresetGameItem(id: "mau_binh", localizationKey: "game_mau_binh", defaultZeroSum: true),
    PresetGameItem(id: "poker", localizationKey: "game_poker", defaultZeroSum: true),
    PresetGameItem(id: "sam", localizationKey: "game_sam", defaultZeroSum: true),
    PresetGameItem(id: "ludo", localizationKey: "game_ludo", defaultZeroSum: false),
    PresetGameItem(id: "monopoly", localizationKey: "game_monopoly", defaultZeroSum: false),
    PresetGameItem(id: "custom", localizationKey: "game_custom", defaultZeroSum: false)
]

struct NewSessionScreen: View {
    @Environment(\.modelContext) private var modelContext
    @Environment(\.colorScheme) private var colorScheme
    
    @Query private var customGames: [CustomGame]
    
    @State private var name: String = ""
    @State private var selectedGameKey: String = "tien_len"
    @State private var customGameName: String = ""
    @State private var playerInput: String = ""
    @State private var players: [String] = []
    @State private var zeroSum: Bool = true
    
    @State private var duplicateName: String? = nil
    @State private var showDuplicateAlert = false
    
    @State private var showMinPlayersAlert = false
    
    let onBack: () -> Void
    let onCreated: (String) -> Void
    
    var body: some View {
        let palette = TallyoTheme.palette(for: colorScheme)
        let canCreate = players.count >= 2
        
        NavigationStack {
            ZStack {
                palette.bg.ignoresSafeArea()
                
                VStack(spacing: 0) {
                    // Custom Navigation Bar
                    HStack {
                        Button(action: onBack) {
                            Image(systemName: "chevron.left")
                                .font(.system(size: 20, weight: .semibold))
                                .foregroundColor(palette.text)
                        }
                        
                        Spacer()
                        
                        Text(NSLocalizedString("new_title", comment: ""))
                            .font(.system(size: 18, weight: .bold))
                            .foregroundColor(palette.text)
                        
                        Spacer()
                        
                        // Empty spacer to center title
                        Color.clear.frame(width: 24, height: 24)
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 12)
                    
                    ScrollView {
                        VStack(alignment: .leading, spacing: 24) {
                            
                            // 1. Game Section
                            NewSessionSection(title: NSLocalizedString("new_game_label", comment: "")) {
                                VStack(alignment: .leading, spacing: 12) {
                                    // Flow row of games
                                    FlowLayout(spacing: 8) {
                                        // Presets (excluding custom option first)
                                        ForEach(PRESET_GAMES.filter { $0.id != "custom" }) { game in
                                            let isSelected = selectedGameKey == game.id
                                            GameChip(
                                                title: NSLocalizedString(game.localizationKey, comment: ""),
                                                isSelected: isSelected,
                                                onTap: {
                                                    selectedGameKey = game.id
                                                    zeroSum = game.defaultZeroSum
                                                }
                                            )
                                        }
                                        
                                        // Saved Custom Games
                                        ForEach(customGames) { customGame in
                                            let key = "custom_saved:\(customGame.name)"
                                            let isSelected = selectedGameKey == key
                                            HStack(spacing: 4) {
                                                Text(customGame.name)
                                                    .font(.system(size: 13, weight: .medium))
                                                    .foregroundColor(isSelected ? palette.primary : palette.text)
                                                
                                                Button(action: {
                                                    modelContext.delete(customGame)
                                                    try? modelContext.save()
                                                    if selectedGameKey == key {
                                                        selectedGameKey = "tien_len"
                                                    }
                                                }) {
                                                    Text("×")
                                                        .font(.system(size: 16, weight: .bold))
                                                        .foregroundColor(isSelected ? palette.primary : palette.textMuted)
                                                        .padding(.leading, 2)
                                                }
                                            }
                                            .padding(.horizontal, 12)
                                            .padding(.vertical, 8)
                                            .background(isSelected ? palette.primaryTintBg : palette.surfaceAlt)
                                            .cornerRadius(999)
                                            .overlay(
                                                RoundedRectangle(cornerRadius: 999)
                                                    .stroke(isSelected ? palette.primary : palette.border, lineWidth: 1)
                                            )
                                            .onTapGesture {
                                                selectedGameKey = key
                                                zeroSum = customGame.defaultZeroSum
                                            }
                                        }
                                        
                                        // Custom option
                                        if let customItem = PRESET_GAMES.first(where: { $0.id == "custom" }) {
                                            let isSelected = selectedGameKey == customItem.id
                                            GameChip(
                                                title: NSLocalizedString(customItem.localizationKey, comment: ""),
                                                isSelected: isSelected,
                                                onTap: {
                                                    selectedGameKey = customItem.id
                                                    zeroSum = customItem.defaultZeroSum
                                                }
                                            )
                                        }
                                    }
                                    
                                    if selectedGameKey == "custom" {
                                        TextField(NSLocalizedString("new_game_custom_placeholder", comment: ""), text: $customGameName)
                                            .padding(12)
                                            .background(palette.surface)
                                            .foregroundColor(palette.text)
                                            .cornerRadius(8)
                                            .overlay(
                                                RoundedRectangle(cornerRadius: 8)
                                                    .stroke(palette.border, lineWidth: 1)
                                            )
                                            .textInputAutocapitalization(.words)
                                    }
                                }
                            }
                            
                            // 2. Session Name Section
                            NewSessionSection(title: NSLocalizedString("new_name_label", comment: "")) {
                                TextField(NSLocalizedString("new_name_placeholder", comment: ""), text: $name)
                                    .padding(12)
                                    .background(palette.surface)
                                    .foregroundColor(palette.text)
                                    .cornerRadius(8)
                                    .overlay(
                                        RoundedRectangle(cornerRadius: 8)
                                            .stroke(palette.border, lineWidth: 1)
                                    )
                                    .textInputAutocapitalization(.sentences)
                            }
                            
                            // 3. Players Section
                            NewSessionSection(
                                title: NSLocalizedString("new_players_label", comment: ""),
                                hint: NSLocalizedString("new_players_hint", comment: "")
                            ) {
                                VStack(alignment: .leading, spacing: 12) {
                                    HStack {
                                        TextField(NSLocalizedString("new_players_placeholder", comment: ""), text: $playerInput)
                                            .padding(12)
                                            .background(palette.surface)
                                            .foregroundColor(palette.text)
                                            .cornerRadius(8)
                                            .overlay(
                                                RoundedRectangle(cornerRadius: 8)
                                                    .stroke(palette.border, lineWidth: 1)
                                            )
                                            .textInputAutocapitalization(.words)
                                            .onSubmit {
                                                addPlayer()
                                            }
                                        
                                        Button(action: addPlayer) {
                                            Text(NSLocalizedString("new_players_add", comment: ""))
                                                .font(.system(size: 14, weight: .bold))
                                                .padding(.horizontal, 16)
                                                .padding(.vertical, 12)
                                                .background(palette.primary)
                                                .foregroundColor(.white)
                                                .cornerRadius(8)
                                        }
                                    }
                                    
                                    if players.isEmpty {
                                        Text(NSLocalizedString("new_players_empty", comment: ""))
                                            .font(.system(size: 11))
                                            .foregroundColor(palette.textMuted)
                                    } else {
                                        FlowLayout(spacing: 8) {
                                            ForEach(players, id: \.self) { p in
                                                HStack(spacing: 6) {
                                                    Text(p)
                                                        .font(.system(size: 13, weight: .medium))
                                                        .foregroundColor(palette.text)
                                                    Text("×")
                                                        .font(.system(size: 16, weight: .bold))
                                                        .foregroundColor(palette.textMuted)
                                                }
                                                .padding(.horizontal, 12)
                                                .padding(.vertical, 8)
                                                .background(palette.surfaceAlt)
                                                .cornerRadius(999)
                                                .overlay(
                                                    RoundedRectangle(cornerRadius: 999)
                                                        .stroke(palette.border, lineWidth: 1)
                                                )
                                                .onTapGesture {
                                                    players.removeAll { $0 == p }
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                            
                            // 4. Scoring Options Section
                            NewSessionSection(title: NSLocalizedString("new_options_label", comment: "")) {
                                Button(action: { zeroSum.toggle() }) {
                                    HStack(alignment: .top, spacing: 12) {
                                        // Checkbox
                                        ZStack {
                                            RoundedRectangle(cornerRadius: 4)
                                                .stroke(zeroSum ? palette.primary : palette.border, lineWidth: 2)
                                                .frame(width: 20, height: 20)
                                            
                                            if zeroSum {
                                                RoundedRectangle(cornerRadius: 4)
                                                    .fill(palette.primary)
                                                    .frame(width: 20, height: 20)
                                                
                                                Image(systemName: "checkmark")
                                                    .font(.system(size: 12, weight: .bold))
                                                    .foregroundColor(.white)
                                            }
                                        }
                                        .padding(.top, 2)
                                        
                                        VStack(alignment: .leading, spacing: 4) {
                                            Text(NSLocalizedString("new_options_zero_sum_title", comment: ""))
                                                .font(.system(size: 16, weight: .semibold))
                                                .foregroundColor(palette.text)
                                            
                                            Text(NSLocalizedString("new_options_zero_sum_desc", comment: ""))
                                                .font(.system(size: 13))
                                                .foregroundColor(palette.textMuted)
                                                .lineSpacing(4)
                                                .multilineTextAlignment(.leading)
                                        }
                                    }
                                    .padding(12)
                                    .background(zeroSum ? palette.primaryTintBg : palette.surface)
                                    .cornerRadius(10)
                                    .overlay(
                                        RoundedRectangle(cornerRadius: 10)
                                            .stroke(zeroSum ? palette.primary : palette.border, lineWidth: 1)
                                    )
                                }
                                .buttonStyle(.plain)
                            }
                            
                            // Submit Button
                            Button(action: submitSession) {
                                Text(NSLocalizedString("new_submit", comment: ""))
                                    .font(.system(size: 16, weight: .bold))
                                    .frame(maxWidth: .infinity)
                                    .padding(.vertical, 14)
                                    .background(canCreate ? palette.primary : palette.primary.opacity(0.4))
                                    .foregroundColor(.white)
                                    .cornerRadius(8)
                            }
                            .disabled(!canCreate)
                            .padding(.top, 16)
                            .padding(.bottom, 48)
                            
                        }
                        .padding(.horizontal, 16)
                    }
                }
            }
            .alert(isPresented: $showDuplicateAlert) {
                Alert(
                    title: Text(NSLocalizedString("new_players_duplicate_title", comment: "")),
                    message: Text(String(format: NSLocalizedString("new_players_duplicate_message", comment: ""), duplicateName ?? "")),
                    dismissButton: .default(Text("OK")) { duplicateName = nil }
                )
            }
            .alert(NSLocalizedString("new_min_players", comment: ""), isPresented: $showMinPlayersAlert) {
                Button("OK", role: .cancel) { }
            }
        }
    }
    
    private func addPlayer() {
        let trimmed = playerInput.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return }
        if players.contains(trimmed) {
            duplicateName = trimmed
            showDuplicateAlert = true
            return
        }
        players.append(trimmed)
        playerInput = ""
    }
    
    private func submitSession() {
        if players.count < 2 {
            showMinPlayersAlert = true
            return
        }
        
        let selectedGameItem = PRESET_GAMES.first { $0.id == selectedGameKey }
        let isCustom = selectedGameKey == "custom" || selectedGameKey.hasPrefix("custom_saved:")
        
        let gameNameToSend: String
        if selectedGameKey == "custom" {
            let customNameTrimmed = customGameName.trimmingCharacters(in: .whitespacesAndNewlines)
            gameNameToSend = customNameTrimmed.isEmpty ? NSLocalizedString("game_custom", comment: "") : customNameTrimmed
        } else if selectedGameKey.hasPrefix("custom_saved:") {
            gameNameToSend = String(selectedGameKey.dropFirst("custom_saved:".count))
        } else {
            gameNameToSend = selectedGameItem.flatMap { NSLocalizedString($0.localizationKey, comment: "") } ?? ""
        }
        
        let defaultSessionName: String
        if isCustom {
            defaultSessionName = gameNameToSend
        } else if let item = selectedGameItem {
            defaultSessionName = NSLocalizedString(item.localizationKey, comment: "")
        } else {
            defaultSessionName = NSLocalizedString("new_session_default_name", comment: "")
        }
        
        let sessionName = name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? defaultSessionName : name.trimmingCharacters(in: .whitespacesAndNewlines)
        
        // 1. If it's a new custom game, save it to the DB so it is in the chip list next time
        if selectedGameKey == "custom" && !customGameName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            let newCustomGameName = customGameName.trimmingCharacters(in: .whitespacesAndNewlines)
            if !customGames.contains(where: { $0.name == newCustomGameName }) {
                let cg = CustomGame(name: newCustomGameName, defaultZeroSum: zeroSum)
                modelContext.insert(cg)
            }
        }
        
        // 2. Create the Session
        let session = Session(name: sessionName, game: gameNameToSend, zeroSum: zeroSum)
        
        // Add players
        for (index, playerName) in players.enumerated() {
            let player = Player(name: playerName, orderIndex: index)
            session.players.append(player)
        }
        
        modelContext.insert(session)
        try? modelContext.save()
        
        onCreated(session.id)
    }
}

struct NewSessionSection<Content: View>: View {
    let title: String
    var hint: String? = nil
    let content: () -> Content
    
    @Environment(\.colorScheme) private var colorScheme
    
    var body: some View {
        let palette = TallyoTheme.palette(for: colorScheme)
        
        VStack(alignment: .leading, spacing: 8) {
            Text(title)
                .font(.system(size: 18, weight: .semibold))
                .foregroundColor(palette.text)
            
            if let hint = hint {
                Text(hint)
                    .font(.system(size: 11))
                    .foregroundColor(palette.textMuted)
                    .padding(.top, -6)
            }
            
            content()
        }
    }
}

struct GameChip: View {
    let title: String
    let isSelected: Bool
    let onTap: () -> Void
    
    @Environment(\.colorScheme) private var colorScheme
    
    var body: some View {
        let palette = TallyoTheme.palette(for: colorScheme)
        
        Text(title)
            .font(.system(size: 13, weight: .medium))
            .foregroundColor(isSelected ? palette.primary : palette.text)
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            .background(isSelected ? palette.primaryTintBg : palette.surfaceAlt)
            .cornerRadius(999)
            .overlay(
                RoundedRectangle(cornerRadius: 999)
                    .stroke(isSelected ? palette.primary : palette.border, lineWidth: 1)
            )
            .onTapGesture(perform: onTap)
    }
}

// Simple FlowLayout implementation for displaying chips nicely
struct FlowLayout: Layout {
    var spacing: CGFloat
    
    init(spacing: CGFloat = 8) {
        self.spacing = spacing
    }
    
    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let bounds = proposal.width ?? 300
        var totalHeight: CGFloat = 0
        var currentWidth: CGFloat = 0
        var currentRowHeight: CGFloat = 0
        
        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if currentWidth + size.width > bounds {
                // wrap
                totalHeight += currentRowHeight + spacing
                currentWidth = size.width + spacing
                currentRowHeight = size.height
            } else {
                currentWidth += size.width + spacing
                currentRowHeight = max(currentRowHeight, size.height)
            }
        }
        totalHeight += currentRowHeight
        return CGSize(width: bounds, height: totalHeight)
    }
    
    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        var x = bounds.minX
        var y = bounds.minY
        var currentRowHeight: CGFloat = 0
        
        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x + size.width > bounds.maxX {
                // wrap
                x = bounds.minX
                y += currentRowHeight + spacing
                currentRowHeight = size.height
            } else {
                currentRowHeight = max(currentRowHeight, size.height)
            }
            view.place(at: CGPoint(x: x, y: y), proposal: ProposedViewSize(size))
            x += size.width + spacing
        }
    }
}
